torch_object_detection_segmenter.py

```diff
@@ -1,31 +1,214 @@
+            batch_size = parameters.get('batch_size', self._default_batch_size)
+
+            # traverse thought all documents which have to be processed
+            flat_docs = docs.traverse_flat(traversal_paths)
+
+            # filter out documents without images
+            filtered_docs = [doc for doc in flat_docs if doc.blob is not None]
+
+            return _batch_generator(filtered_docs, batch_size)
+
         if not docs:
             return
-        for doc in docs:
-            assert doc.blob.shape[0] == 256
-            doc.embedding = np.random.rand(self._dim)
\ No newline at end of file
+
+
+        batch = _get_input_data(docs, parameters) # a generator of batches of docs
+
+        for i, docs_batch in enumerate(batch):
+            # the blob dimension of imgs/cars.jpg at this point is (2, 681, 1264, 3)
+            # "Ensure the color channel axis is the default axis." i.e. c comes first
+            # e.g. (h,w,c) -> (c,h,w) / (b,h,w,c) -> (b,c,h,w)
+            blob_batch = [_move_channel_axis(d.blob, self.channel_axis,
+                                       self._default_channel_axis) for d in docs_batch]
+
+            all_predictions = self._predict(blob_batch)
+
+            for blob, predictions in zip(blob_batch, all_predictions):
+                bboxes = predictions['boxes'].detach()
+                scores = predictions['scores'].detach()
+                labels = predictions['labels']
+                if self.on_gpu:
+                    bboxes = bboxes.cpu()
+                    scores = scores.cpu()
+                    labels = labels.cpu()
+                img = _load_image(blob * 255, self._default_channel_axis)
+
+                for bbox, score, label in zip(bboxes.numpy(), scores.numpy(), labels.numpy()):
+                    if score >= self.confidence_threshold:
+                        x0, y0, x1, y1 = bbox
+                        # note that tensors are [H, W] while PIL Images are [W, H]
+                        top, left = int(y0), int(x0)
+                        # target size must be (h, w)
+                        target_size = (int(y1) - int(y0), int(x1) - int(x0))
+                        # at this point, raw_img has the channel axis at the default tensor one
+                        _img, top, left = _crop_image(img, target_size=target_size, top=top, left=left, how='precise')
+                        _img = _move_channel_axis(np.asarray(_img).astype('float32'), -1, self.channel_axis)
+                        label_name = self.label_name_map[label]
+                        self.logger.debug(
+                            f'detected {label_name} with confidence {score} at position {(top, left)} and size {target_size}')
+
+                        # a chunk is created for each of the objects detected for each image
+                        d = Document(offset=0, weight=1., blob = _img, location=[top, left], tags={'label': label_name})
+                        docs[i].chunks.append(d)
```

Comment:
This i is the batch index not the doc index. U need to add docs_batch to fhe zip so that u access the Document

Reply:
makes so much sense now